### ISSUE TITLE
Chemu Updates

### ISSUE THREAD

**Issue by gustycooper** [gustycooper/charm#1]
The following list shows potential capability updates to `chemu`.

1. The functions `printres` and `update_display` are defined in `chemu.c`. They are interface functions that other Chemu modules use; however, they are not declared in a `chemu.h` file. The `.c` modules that use `printres` and `update_display` declare function prototypes. Create a `chemu.h` file with the function prototypes that is included by the `.c` modules that use the functions. When `chemu.h` is created, you must update the depends section `setup.py` to include `chemu.h`.
2. Chemu's `step` function in `cpu.c` implements of add with carry (`adc`) and subtract with carry (`sbc`) incorrectly. The implementation is a simple add and subtract. They do not use the carry bit. This error does not cause any problems in the OS code and code that students have written; however, it would be nice to have correct implementations of `adc` and `sbc`. Update the code to correctly implement `adc` and `sbc`.
3. Users writing code for `os.s` and `chaos.s` can call `scanf`. The Chemu functions `chemuioi` and `chemuscanf` in `cpu.c` implement `scanf` by printing "Waiting on input from scanf:" in the display. The next input typed in the command window is read and processed by `scanf`. At this point the assembly program is stopped. Then the user must enter `s` or `run` to continue running. This approach provides a `scanf` that works, but I am sure this approach to `scanf` can be improved.
4. The function `updateregwin` in chemu.c displays the registers for `chemun`. Every time a step command (`s`) is entered, the registers are updated. Most of the registers are unchanged, but one or two of them are changed. Implement an algorithm that colors registers that change as a result of a step (`s`) command, which would allow users to more easily see changed registers. NOTE: `charmweb` colors changed registers.
5. The function `pipeline` in `cpu.c` keeps a history of 11 instructions - 5 that have executed, the instruction about to execute, and the next 5 instructions. When the instruction about to execute is a branch instruction, `pipeline` displays both portions of the branch - the five instructions immediately after the branch and the five instructions instructions that are the destination of the branch. `pipeline` gets the address of the instructions of branch destination from the bottom 20 bits of the instruction. There are branch instructions (for example `b [reg]`) where the destination address is in a register - not the bottom 20 bits. Fix this to be correct for all branch instructions.
6. Chemu supports two breakpoints with the `b` and `b1` commands. The two breakpoints are stored in the global variables `breakpoint` and `breakpoint1`, which are defined in `cpu.c`. The function `do_command` in `command.c` sets the variables when a user enters a `b` or `b1` command. The function `step` in `cpu.c` examines the variables to determine if a breakpoint has been hit. Two breakpoints have been sufficient, but the design could be improved with an array to hold the breakpoints instead of the two variables. This design would allow additional breakpoints to be more easily added.
7. Both `os.s` and `chaos.s` have a `printf` function that is similar to C's printf. The `os.s` and `chaos.s` `printf` has a string as its first parameter. The string  may contain one or two `%` formats - for example `%d`. The parameters following the format string match the `%` formats. For example, an integer parameter matches `%d` and the address of a string matches `%s`. If you have one `%` format in the format string, it can be `%d`, `%f`, or `%s`. If you have two `%` formats in the format string, they can be `%d %d`, `%d %s`, `%s %d`, or `%s %s`. Update the two `%` format strings to accept `%f` and any combinations of `%d`, `%s`, and `%f`. The function `chemuio` in `cpu.c` processes `printf`.
8. `chaos.s` manages processes with a `struct proc`. The information about processes are displayed when a user enters a `ps` command. The function `do_command` of `command.c` processes the `ps` command. `command.c` defines the `struct proc` using a collection of `#define` statements. There is probably a better way to do this. A couple of examples of these defines are the following. 
    a. `#define PROC_SIZE 64` - the size of `chaos.s` `struct proc`
    b. `#define PROC_PID 0` - the offset of the proc's PID in `struct proc`
9. `chaos.s` uses two addresses to communicate with `chemu` - eight bytes beginning at address 0 contains the name of the proc in user most and four bytes at address `0xfff0` has a kernel function marker, which is an integer indicating the kernel function that was just called. `chemu.c` uses these addresses. When in user mode, the name of the proc is displayed next to the PC in the instruction window. When running in slow motion, the function name in the array `char *os_happenings[]` is displayed using the integer as an index into `os_happenings`. This processing occurs in the slow motion command processing of `do_cmd` of `command.c`.
10. Add a memory management unit (MMU) to Chemu that processes page tables. Must think through the format of the virtual memory address. Probably a two-level page table. After adding this, update Charm OS to support virtual memory.
